pkg/sql/drop_test.go

```diff
@@ -916,6 +916,133 @@ CREATE TABLE t.kv (k CHAR PRIMARY KEY, v CHAR);
 
 }
 
+func TestDropTableUnsplitManuallySplitRanges(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	params, _ := tests.CreateTestServerParams()
+
+	defer gcjob.SetSmallMaxGCIntervalForTest()()
+
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(context.Background())
+	ctx := context.Background()
+
+	// Disable strict GC TTL enforcement because we're going to shove a zero-value
+	// TTL into the system with AddImmediateGCZoneConfig.
+	defer sqltestutils.DisableGCTTLStrictEnforcement(t, sqlDB)()
+
+	const numRows = 2*row.TableTruncateChunkSize + 1
+	const numKeys = 3 * numRows
+
+	tableName := "test1"
+	if err := tests.CreateKVTable(sqlDB, tableName, numRows); err != nil {
+		t.Fatal(err)
+	}
+
+	var tableDesc = catalogkv.TestingGetTableDescriptor(kvDB, keys.SystemSQLCodec, "t", tableName)
+
+	nameKey := catalogkeys.MakePublicObjectNameKey(keys.SystemSQLCodec, keys.MinNonPredefinedUserDescID, tableName)
+	gr, err := kvDB.Get(ctx, nameKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !gr.Exists() {
+		t.Fatalf("Name entry %q does not exist", nameKey)
+	}
+
+	tableSpan := tableDesc.TableSpan(keys.SystemSQLCodec)
+	metaStartKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.Key))
+	metaEndKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.EndKey))
+
+	tests.CheckKeyCount(t, kvDB, tableSpan, numKeys)
+
+	// The closure checks whether there is any range has sticky bit
+	// hasManualSplitRange==true: fails when no ranges have sticky bit
+	// hasManualSplitRange==false: fails when any range has sticky bit
+	checkManuallySplitRange := func(hasManualSplitRange bool) {
+		ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+		if err != nil {
+			t.Fatal(err)
+		}
+		require.NotEmpty(t, ranges)
+		hasManuallySplitRange := false
+		for _, r := range ranges {
+			var desc roachpb.RangeDescriptor
+			if err := r.ValueProto(&desc); err != nil {
+				t.Fatal(err)
+			}
+			hasManuallySplitRange = hasManuallySplitRange || !desc.GetStickyBit().IsEmpty()
+		}
+		require.Equal(t, hasManuallySplitRange, hasManualSplitRange)
+	}
+
+	// assert that all ranges are not manually split
+	checkManuallySplitRange(false)
+
+	// split the last range
+	ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lastRange := ranges[len(ranges)-1]
+	var lastRangeDesc roachpb.RangeDescriptor
+	if err := lastRange.ValueProto(&lastRangeDesc); err != nil {
+		t.Fatal(err)
+	}
+	splitKey := tests.GetMidKeyInSpan(t, kvDB, lastRangeDesc.StartKey, lastRangeDesc.EndKey)
+	splitKey, err = keys.EnsureSafeSplitKey(splitKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := kvDB.AdminSplit(ctx, splitKey, hlc.MaxTimestamp); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := sqlDB.Exec(fmt.Sprintf(`DROP TABLE t.%s`, tableName)); err != nil {
+		t.Fatal(err)
+	}
+
+	// Data hasn't been GC-ed.
```

Comment:
It's worth mentioning in this comment that the reason it hasn't been GC-ed is that the GC TTL is still set to its original high value (I think the default is 25h).